### PR TITLE
add encoding parameter to creating tar entry

### PR DESCRIPTION
**Pull request by itn3000**
This PR is similar to #182, but this PR adds System.Text.Encoding to TarEntry related API.
default is same as current master behavior(omit upper byte) because more faster than using encoding if tar archive assured having ASCII filename only.

_I certify that I own, and have sufficient rights to contribute, all source code and related material intended to be compiled or integrated with the source code for the SharpZipLib open source product (the "Contribution"). My Contribution is licensed under the MIT License._

### CODE REVIEW

**Comment by itn3000** [comment]
Codacy says "`TarInputStream.CreateEntry(byte[] headerBuffer, Encoding enc)` should be static method".
But other `TarInputStream.CreateEntry` functions are not static, should I `CreateEntry(byte[] headerBuffer, Encoding enc)` make static?

**Comment by piksel** [comment]
The problem right now with Codacy is that the current code right now is way beyond bad.  
I think it's still better to be consistent here than to adhere to Codacy since it's better to do a full cleanup in another commit of all the functions.

**Comment by piksel** [comment]
As stated in #182:
The specification suggests using extended headers for encodings other than ASCII, but GNU Tar writes the UTF8-encoded bytes as the Name so I guess we should too.

This should be fine.

File: src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs [hunk 1]
```diff
@@ -252,23 +287,23 @@ public void PutNextEntry(TarEntry entry)
...
 				buffer.WriteBlock(blockBuffer);  // Add special long filename header block
 
 				int nameCharIndex = 0;
 
-				while (nameCharIndex < entry.TarHeader.Name.Length + 1 /* we've allocated one for the null char, now we must make sure it gets written out */)
+				while (nameCharIndex < namelen + 1 /* we've allocated one for the null char, now we must make sure it gets written out */)
 				{
 					Array.Clear(blockBuffer, 0, blockBuffer.Length);
-					TarHeader.GetAsciiBytes(entry.TarHeader.Name, nameCharIndex, this.blockBuffer, 0, TarBuffer.BlockSize); // This func handles OK the extra char out of string length
+					TarHeader.GetAsciiBytes(entry.TarHeader.Name, nameCharIndex, this.blockBuffer, 0, TarBuffer.BlockSize, encoding); // This func handles OK the extra char out of string length
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs:300]: This should probably be renamed?

File: src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs [hunk 2]
```diff
@@ -475,6 +510,11 @@ private void WriteEofBlock()
 		/// </summary>
 		protected Stream outputStream;
 
+		/// <summary>
+		/// name encoding
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs:514]:
```suggestion
		/// The <see cref="Encoding"/> used for the Name fields
```

File: test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs [hunk 3]
```diff
@@ -831,7 +832,38 @@ public void ExtractingCorruptTarShouldntLeakFiles()
...
+		[Category("Tar")]
+		public void StreamWithJapaneseName(int length)
+		{
+			// U+3042 is Japanese Hiragana
+			// https://unicode.org/charts/PDF/U3040.pdf
+			var entryName = new string((char)0x3042, length);
+			var data = new byte[32];
+			using(var memoryStream = new MemoryStream())
+			{
+				using(var tarOutput = new TarOutputStream(memoryStream, System.Text.Encoding.UTF8))
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs:851]:
```suggestion
				using(var tarOutput = new TarOutputStream(memoryStream, Encoding.UTF8))
```

**Comment by itn3000** [comment]
Thank you for reviewing, I'd like to reflect the review results.

File: src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs (hunk 1 above)
Comment (itn3000) [reply src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs:300]: Indeed, it should be renamed to `nameEncoding`

**Comment by piksel** [comment]
Regarding documentation:
Perhaps something short, like "or null for ASCII"?

Regarding Encoding.UTF8: 
That's odd? The documentation for that property explicitly warns about how it emits them:
https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.utf8?view=netframework-4.8#remarks

File: src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs (hunk 1 above)
Comment (piksel) [reply src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs:300]: Sorry, I meant the "GetAsciiBytes" method. The name is not really applicable when another encoding is used.

**Comment by itn3000** [comment]
I've almost done with requested changes, please review it.

File: src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs [hunk 4]
```diff
@@ -139,15 +167,15 @@ public static TarArchive CreateInputTarArchive(Stream inputStream, int blockFact
...
 
-			return new TarArchive(new TarInputStream(inputStream, blockFactor));
+			return new TarArchive(new TarInputStream(inputStream, blockFactor, nameEncoding));
 		}
-
 		/// <summary>
 		/// Create a TarArchive for writing to, using the default blocking factor
 		/// </summary>
 		/// <param name="outputStream">The <see cref="Stream"/> to write to</param>
+		/// <param name="nameEncoding"></param>
```
Comment (Numpsy) [inline src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs:176]: not trying to do a real review (not up to speed with the changes/requirements), but fwiw - empty doc comment for the new param?

File: src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs [hunk 5]
```diff
@@ -175,6 +212,17 @@ public static TarArchive CreateOutputTarArchive(Stream outputStream)
...
 		public static TarArchive CreateOutputTarArchive(Stream outputStream, int blockFactor)
+		{
+			return CreateOutputTarArchive(outputStream, blockFactor, null);
+		}
+		/// <summary>
+		/// Create a <see cref="TarArchive">tar archive</see> for writing.
+		/// </summary>
+		/// <param name="outputStream">The stream to write to</param>
+		/// <param name="blockFactor">The blocking factor to use for buffering.</param>
+		/// <param name="nameEncoding"></param>
```
Comment (Numpsy) [inline src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs:223]: ditto

File: src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs (hunk 4 above)
Comment (itn3000) [reply src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs]: There were comfirmation omission. I inserted comment.

File: src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs (hunk 5 above)
Comment (itn3000) [reply src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs]: There were comfirmation omission. I inserted comment.

File: src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs (hunk 4 above)
Comment (piksel) [reply src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs]: Yeah, I think that the variable name here does enough to explain the purpose. I see little to no gain in writing descriptions like "The Encoding used for the name".

Comment (itn3000) [reply src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs]: Sorry for late response.
Should I leave this parameter comment empty?

Comment (piksel) [reply src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs]: It's fine

**Pull request merged**